### PR TITLE
A bunch of minor modifications

### LINKED ISSUE #10168: Multiple conditions with commas at CV_Assert
##### System information (version)
- OpenCV => 3.3.1
- Operating System / Platform => Ubuntu 16.04
- Compiler => /usr/bin/c++ (ver 5.4.0)

CV_Assert with multiple conditions defined as separate code lines.

##### Steps to reproduce

```cpp
if (true)
  CV_Assert(1 != 2);
else
  CV_Assert(1 == 2, 2 == 3);
```
produces
```
error: (-215) 2 == 3
```

### PR DESCRIPTION
**Pull request by mshabunin**
resolves #10168

### This pullrequest changes

- fixes _CV_Assert_ macro for static analyzers
- disabled one warning in protobuf library
- added [ThinLTO](https://clang.llvm.org/docs/ThinLTO.html) compilation option
- some minor issues

```
allow_multiple_commits=1
```

### CODE REVIEW

File: modules/core/include/opencv2/core/base.hpp [hunk 1]
```diff
@@ -445,17 +449,16 @@ for example:
...
-*/
+/** same as CV_Error(code,msg), but does not return */
+#define CV_ErrorNoReturn( code, msg ) cv::errorNoReturn( code, msg, CV_Func, __FILE__, __LINE__ )
 
-#define CV_VA_NUM_ARGS_HELPER(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10, N, ...)    N
-#define CV_VA_NUM_ARGS(...)      CV_VA_NUM_ARGS_HELPER(__VA_ARGS__, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+/** same as CV_Error_(code,args), but does not return */
+#define CV_ErrorNoReturn_( code, args ) cv::errorNoReturn( code, cv::format args, CV_Func, __FILE__, __LINE__ )
 
 #define CV_Assert_1( expr ) if(!!(expr)) ; else cv::error( cv::Error::StsAssert, #expr, CV_Func, __FILE__, __LINE__ )
```
Comment (csukuangfj) [inline modules/core/include/opencv2/core/base.hpp]: Should `CV_Assert_1` to 10 be wrapped inside `do {...} while(0)`

Comment (alalek) [reply modules/core/include/opencv2/core/base.hpp]: It is from here: #10168

File: modules/calib3d/src/fisheye.cpp [hunk 2]
```diff
@@ -762,7 +762,7 @@ double cv::fisheye::calibrate(InputArrayOfArrays objectPoints, InputArrayOfArray
 
 
     //-------------------------------Optimization
-    for(int iter = 0; iter <= std::numeric_limits<int>::max(); ++iter)
+    for(int iter = 0; iter < std::numeric_limits<int>::max(); ++iter)
```
Comment (alalek) [inline modules/calib3d/src/fisheye.cpp]:
```
for(int iter = 0; ; ++iter)
```
?

File: modules/core/src/dxt.cpp [hunk 3]
```diff
@@ -2162,9 +2162,7 @@ static bool ocl_dft(InputArray _src, OutputArray _dst, int flags, int nonzero_ro
     // if output format is not specified
     if (complex_output + real_output == 0)
     {
-        if (real_input)
-            real_output = 1;
-        else
+        if (!real_input)
             complex_output = 1;
```
Comment (alalek) [inline modules/core/src/dxt.cpp]: Looks like this code is not equivalent.

Does this warning is "set but not used"?
My suggestion is to drop "real_input" and "real_output" flags (replace them to "!complex_input" and "!complex_output") - we don't support [quaternions](http://ieeexplore.ieee.org/document/615636/) or similar here.

File: modules/core/include/opencv2/core/base.hpp (hunk 1 above)
Comment (mshabunin) [reply modules/core/include/opencv2/core/base.hpp]: I've added mentioned construction to the `CV_Assert` macro.

File: modules/core/src/dxt.cpp (hunk 3 above)
Comment (mshabunin) [reply modules/core/src/dxt.cpp]: I've refactored this part of function to be more readable.

**Comment by mshabunin** [comment]
Rebased on top of current _master_ branch.

**Pull request merged**